src/vs/editor/contrib/format/formatActions.ts

```diff
@@ -263,11 +263,13 @@ class FormatSelectionAction extends EditorAction {
 		}
 		const instaService = accessor.get(IInstantiationService);
 		const model = editor.getModel();
-		let range: Range = editor.getSelection();
-		if (range.isEmpty()) {
-			range = new Range(range.startLineNumber, 1, range.startLineNumber, model.getLineMaxColumn(range.startLineNumber));
-		}
-		await instaService.invokeFunction(formatDocumentRangeWithSelectedProvider, editor, range, FormattingMode.Explicit, CancellationToken.None);
+		await editor.getSelections().reduce(async (prev, range: Range) => {
+			await prev;
+			if (range.isEmpty()) {
+				range = new Range(range.startLineNumber, 1, range.startLineNumber, model.getLineMaxColumn(range.startLineNumber));
+			}
+			await instaService.invokeFunction(formatDocumentRangeWithSelectedProvider, editor, range, FormattingMode.Explicit, CancellationToken.None);
+		}, Promise.resolve());
```

Comment:
Formatting one will fire `onDidChangeModelContent` event, and another will be cancelled.